sqlx-core/src/sqlite/connection.rs

```diff
@@ -0,0 +1,159 @@
+use core::ptr::{null, null_mut, NonNull};
+
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::ffi::CString;
+
+use futures_core::future::BoxFuture;
+use futures_util::future;
+use libsqlite3_sys::{
+    sqlite3, sqlite3_close, sqlite3_extended_result_codes, sqlite3_open_v2, SQLITE_OK,
+    SQLITE_OPEN_CREATE, SQLITE_OPEN_NOMUTEX, SQLITE_OPEN_READWRITE, SQLITE_OPEN_SHAREDCACHE,
+};
+
+use crate::connection::{Connect, Connection};
+use crate::executor::Executor;
+use crate::sqlite::statement::SqliteStatement;
+use crate::sqlite::worker::Worker;
+use crate::sqlite::SqliteError;
+use crate::url::Url;
+
+#[derive(Clone, Copy)]
+pub(super) struct SqliteConnectionHandle(NonNull<sqlite3>);
+
+pub struct SqliteConnection {
+    pub(super) worker: Worker,
+    pub(super) handle: SqliteConnectionHandle,
+    // Storage of the most recently prepared, non-persistent statement
+    pub(super) statement: Option<SqliteStatement>,
+    // Storage of persistent statements
+    pub(super) statements: Vec<SqliteStatement>,
+    pub(super) statement_by_query: HashMap<String, usize>,
+}
+
+// SAFE: A sqlite3 handle is safe to access from multiple threads provided
+//       that only one thread access it at a time. Or in other words,
+//       the same guarantees that [Sync] requires. This is upheld as long
```

Comment:
(pedantry) wouldn't this be `Send`, not `Sync`? `Sync` means multiple threads can access simultaneously.

Reply:
> [...] is safe to access from multiple threads provided that only one thread access it at a time [...]

That is `Sync` with a mutable receiver `&mut self`. I updated the notes to make it clear that we need to ensure use of `&mut self` to uphold this.

Reply:
Nevermind. We don't even need `Sync` for `SqliteConnection`.